jormungandr/src/blockchain/storage.rs

```diff
@@ -156,17 +157,83 @@ enum StreamingError {
     ),
 }
 
+enum Query {
+    PutBlock(Block, oneshot::Sender<Result<(), StorageError>>),
+    PutTag(
+        String,
+        HeaderHash,
+        oneshot::Sender<Result<(), StorageError>>,
+    ),
+}
+
+pub struct StorageSyncQueryExecutor {
+    queries: mpsc::Receiver<Query>,
+    pool: Pool<ConnectionManager>,
+}
+
+impl StorageSyncQueryExecutor {
+    pub async fn start(self) {
+        let StorageSyncQueryExecutor { mut queries, pool } = self;
+
+        let mut queries = stream::poll_fn(move |cx| {
+            let mut queries_vec = Vec::new();
+            loop {
+                let queries_queue = Pin::new(&mut queries);
+                match queries_queue.poll_next(cx) {
+                    Poll::Ready(Some(query)) => queries_vec.push(query),
+                    Poll::Ready(None) => return Poll::Ready(None),
```

Comment:
potentially broken: there may be some blocks in `queries_queue` which should be returned